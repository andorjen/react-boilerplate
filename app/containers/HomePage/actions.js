/**
 * Actions for HomePage
 */
import {
  LOAD_ALL_CONTENTS,
  LOAD_ALL_CONTENTS_SUCCESS,
  LOAD_ALL_CONTENTS_ERROR,
} from './constants';

/** returns an action with type of LOAD_ALL_CONTENTS */
function requestAllContents() {
  return {
    type: LOAD_ALL_CONTENTS,
  };
}

/** returns an action with type LOAD_ALL_CONTENTS_SUCCESS, and payload of data */
function requestAllContentsSuccess(data) {
  return {
    type: LOAD_ALL_CONTENTS_SUCCESS,
    payload: data,
  };
}

/** returns an action with type LOAD_ALL_CONTENTS_ERROR, and payload of err */
function requestAllContentsError(err) {
  return {
    type: LOAD_ALL_CONTENTS_ERROR,
    payload: err,
  };
}

export {
  requestAllContents,
  requestAllContentsSuccess,
  requestAllContentsError,
};
