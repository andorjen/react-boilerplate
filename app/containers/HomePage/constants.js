const BACKEND_HOST = 'localhost:3000/api/contents';
const LOAD_ALL_CONTENTS = 'LOAD_ALL_CONTENTS';
const LOAD_ALL_CONTENTS_SUCCESS = 'LOAD_ALL_CONTENTS_SUCCESS';
const LOAD_ALL_CONTENTS_ERROR = 'LOAD_ALL_CONTENTS_ERROR';

export {
  BACKEND_HOST,
  LOAD_ALL_CONTENTS,
  LOAD_ALL_CONTENTS_SUCCESS,
  LOAD_ALL_CONTENTS_ERROR,
};
