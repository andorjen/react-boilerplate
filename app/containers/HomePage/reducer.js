import {
  LOAD_ALL_CONTENTS,
  LOAD_ALL_CONTENTS_ERROR,
  LOAD_ALL_CONTENTS_SUCCESS,
} from './constants';

// The initial state of the HomePage
export const initialState = {
  isLoading: false,
  contents: [],
};

function homeReducer(state = initialState, { type, payload }) {
  switch (type) {
    case LOAD_ALL_CONTENTS:
      return { ...state, isLoading: true };
    case LOAD_ALL_CONTENTS_SUCCESS:
      return { ...state, isLoading: false, contents: payload.contents };
    case LOAD_ALL_CONTENTS_ERROR:
      return { ...state, isLoading: false, error: payload.error };
    default:
      return state;
  }
}

export default homeReducer;
