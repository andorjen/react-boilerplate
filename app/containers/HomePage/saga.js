import { call, put, takeLatest, delay } from 'redux-saga/effects';
import axios from 'axios';

import {
  LOAD_ALL_CONTENTS,
  LOAD_ALL_CONTENTS_ERROR,
  LOAD_ALL_CONTENTS_SUCCESS,
  BACKEND_URL,
} from './constants';

/** Make Api Call to backend to get all data */
function* fetchAllData() {
  try {
    yield delay(500); // mimick the wait time for api call here, not necessary
    const res = yield call(axios.get, BACKEND_URL);

    // if success, pass response data in reversed order to reducer
    yield put({
      type: LOAD_ALL_CONTENTS_SUCCESS,
      payload: { contents: res.data.contents.reverse() },
    });
  } catch (err) {
    // if error, pass on error to reducer
    yield put({
      type: LOAD_ALL_CONTENTS_ERROR,
      payload: { error: err.response.data.error },
    });
  }
}

/** perform fetchALlData on the latest action of LOAD_ALL_CONTENTS */
function* homeSaga() {
  yield takeLatest(LOAD_ALL_CONTENTS, fetchAllData);
}

export default homeSaga;
