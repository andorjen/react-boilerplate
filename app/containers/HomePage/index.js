/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 *
 */

import React, { useEffect, memo } from 'react';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import { compose } from 'redux';

import { useInjectReducer } from 'utils/injectReducer';
import { useInjectSaga } from 'utils/injectSaga';
import {
  makeSelectContents,
  makeSelectLoading,
  makeSelectError,
} from './selectors';

import CardList from '../../components/CardList';
import NavBar from '../../components/NavBar';
import LoadingIndicator from '../../components/LoadingIndicator';

import { requestAllContents } from './actions';

import homeReducer from './reducer';
import homeSaga from './saga';

function HomePage(props) {
  useInjectReducer({ key: 'homeReducer', reducer: homeReducer });
  useInjectSaga({ key: 'homeSaga', saga: homeSaga });
  useEffect(() => {
    props.getAllData();
  }, []);

  return (
    <div>
      <NavBar />
      {props.isLoading && <LoadingIndicator />}
      {props.isLoading && <h2>Loading...</h2>}
      {props.error && <h3>{props.error.message}</h3>}

      {!props.error && props.contents.length > 0 && (
        <CardList
          contents={props.contents}
          onClick={() => {
            console.log('clicked');
          }}
        />
      )}
    </div>
  );
}

HomePage.propTypes = {
  isLoading: PropTypes.bool,
  error: PropTypes.object,
  contents: PropTypes.array,
  getAllData: PropTypes.func,
};
const mapStateToProps = createStructuredSelector({
  isLoading: makeSelectLoading(),
  error: makeSelectError(),
  contents: makeSelectContents(),
});

function mapDispatchToProps(dispatch) {
  return {
    getAllData: () => {
      dispatch(requestAllContents());
    },
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(HomePage);
