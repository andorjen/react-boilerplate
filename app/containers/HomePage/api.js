import axios from 'axios';
import { BACKEND_URL } from './constants';

async function fetchAllContents() {
  const res = await axios.get(BACKEND_URL);
  return res.data;
}

export default fetchAllContents;
