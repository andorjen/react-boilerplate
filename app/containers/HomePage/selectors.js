import { createSelector } from 'reselect';
import { initialState } from './reducer';
const selectRouter = state => state.router;
const selectHome = state => state.homeReducer || initialState;

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

const makeSelectLoading = () =>
  createSelector(
    selectHome,
    globalState => globalState.isLoading,
  );

const makeSelectError = () =>
  createSelector(
    selectHome,
    globalState => globalState.error,
  );

const makeSelectContents = () =>
  createSelector(
    selectHome,
    globalState => globalState.contents,
  );

const makeSelectNeedsLoading = () =>
  createSelector(
    selectHome,
    globalState => globalState.needsLoading,
  );

export {
  selectHome,
  makeSelectLocation,
  makeSelectLoading,
  makeSelectError,
  makeSelectContents,
  makeSelectNeedsLoading,
};
