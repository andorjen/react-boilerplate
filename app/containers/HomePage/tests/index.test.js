import React from 'react';
import { render } from 'react-testing-library';
// import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import history from 'utils/history';
import HomePage from '../index';
import configureStore from '../../../configureStore';

// Create redux store with history
const initialState = {};
const store = configureStore(initialState, history);

describe('<HomePage />', () => {
  it('should render and match the snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <HomePage />
        </MemoryRouter>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
