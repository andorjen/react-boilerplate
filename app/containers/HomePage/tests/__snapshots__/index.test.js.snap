// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<div>
  <div>
    <nav
      class="NavBar__Nav-ar0vx-0 haeefk"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
        style="background-color: rgba(140, 71, 245, 0.4); font-weight: bold;"
      >
        Home
      </a>
      <a
        href="/add"
      >
        Add A Card
      </a>
    </nav>
    <div
      class="HomePage__HomeHeader-i81cdn-0 exSvdz"
    >
      <div
        class="HomePage__PaddedWrapper-i81cdn-1 kFNdm"
      >
        <div
          class="Wrapper-sc-12uw37d-0 bTWIVF"
        >
          <div
            class="Circle__CirclePrimitive-ww56dy-0 enQIMS"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 jCxmwS"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 dyFXB"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 ikwUeP"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 hwlQRz"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 jjXxyb"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 jVExkJ"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 haCVIz"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 ghhiLZ"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 bCbqdt"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 hZPQMA"
          />
          <div
            class="Circle__CirclePrimitive-ww56dy-0 fsGOmP"
          />
        </div>
      </div>
      <div
        class="HomePage__PaddedWrapper-i81cdn-1 kFNdm"
      >
        <h2>
          Loading...
        </h2>
      </div>
    </div>
  </div>
</div>
`;
